File: visualstudio-extension/src/Atec.SourceTrace.Core/EditorContextResolver.cs
using System;
using System.Collections.Generic;

namespace Atec.SourceTrace.Core;

/// <summary>Port of vscode-extension editorContext.ts resolveEditorContext.</summary>
public static class EditorContextResolver
{
    public enum SourceMode
    {
        Selection,
        SelectionSymbol,
        CursorWord,
        RecentSelectionFallback,
        None
    }

    public sealed class Resolved
    {
        public string SelectedText { get; }
        public SourceMode SourceMode { get; }
        public string? DetectedSymbol { get; }

        public Resolved(string selectedText, SourceMode sourceMode, string? detectedSymbol)
        {
            SelectedText = selectedText;
            SourceMode = sourceMode;
            DetectedSymbol = detectedSymbol;
        }

        public string SourceModeWire => SourceMode switch
        {
            SourceMode.Selection => "selection",
            SourceMode.SelectionSymbol => "selection_symbol",
            SourceMode.CursorWord => "cursor_word",
            SourceMode.RecentSelectionFallback => "recent_selection_fallback",
            _ => "none"
        };
    }

    public static Resolved? Resolve(
        string? immediateSelectionText,
        string? recentSelectionText,
        string? cursorWord,
        string? currentLineText)
    {
        var immediate = Trim(immediateSelectionText);
        if (immediate.Length > 0)
        {
            return FromText(immediate, SourceMode.Selection);
        }

        var recent = Trim(recentSelectionText);
        if (recent.Length > 0)
        {
            var r = FromText(recent, SourceMode.RecentSelectionFallback);
            return new Resolved(r.SelectedText, SourceMode.RecentSelectionFallback, r.DetectedSymbol);
        }

        var word = Trim(cursorWord);
        if (word.Length > 0 && SymbolExtractor.ExtractDetectedSymbol(word) != null)
        {
            return new Resolved(word, SourceMode.CursorWord, word);
        }

        var line = Trim(currentLineText);
        if (line.Length > 0)
        {
            var symbol = SymbolExtractor.ExtractDetectedSymbol(line);
            if (symbol != null)
            {
                return new Resolved(line, SourceMode.CursorWord, symbol);
            }
        }

        return null;
    }

    private static Resolved FromText(string text, SourceMode baseMode)
    {
        var trimmed = text.Trim();
        var symbol = SymbolExtractor.ExtractDetectedSymbol(trimmed);
        var isSingleSymbol = symbol != null && trimmed.Equals(symbol, StringComparison.Ordinal);
        return new Resolved(
            trimmed,
            isSingleSymbol ? SourceMode.SelectionSymbol : baseMode,
            symbol);
    }

    private static string Trim(string? s) => (s ?? string.Empty).Trim();
}

File: eclipse-plugin/com.atec.sourcetrace.eclipse/src/com/atec/sourcetrace/eclipse/core/EditorContextResolver.java
package com.atec.sourcetrace.eclipse.core;

/**
 * Port of vscode-extension {@code editorContext.ts} resolveEditorContext.
 */
public final class EditorContextResolver {

	public enum SourceMode {
		SELECTION("selection"),
		SELECTION_SYMBOL("selection_symbol"),
		CURSOR_WORD("cursor_word"),
		RECENT_SELECTION_FALLBACK("recent_selection_fallback"),
		NONE("none");

		public final String wire;

		SourceMode(String wire) {
			this.wire = wire;
		}
	}

	public static final class Resolved {
		public final String selectedText;
		public final SourceMode sourceMode;
		public final String detectedSymbol;

		public Resolved(String selectedText, SourceMode sourceMode, String detectedSymbol) {
			this.selectedText = selectedText;
			this.sourceMode = sourceMode;
			this.detectedSymbol = detectedSymbol;
		}
	}

	private EditorContextResolver() {
	}

	public static Resolved resolve(
			String immediateSelectionText,
			String recentSelectionText,
			String cursorWord,
			String currentLineText) {
		String immediate = trim(immediateSelectionText);
		if (!immediate.isEmpty()) {
			return fromText(immediate, SourceMode.SELECTION);
		}
		String recent = trim(recentSelectionText);
		if (!recent.isEmpty()) {
			Resolved r = fromText(recent, SourceMode.RECENT_SELECTION_FALLBACK);
			return new Resolved(r.selectedText, SourceMode.RECENT_SELECTION_FALLBACK, r.detectedSymbol);
		}
		String word = trim(cursorWord);
		if (!word.isEmpty() && SymbolExtractor.extractDetectedSymbol(word) != null) {
			return new Resolved(word, SourceMode.CURSOR_WORD, word);
		}
		String line = trim(currentLineText);
		if (!line.isEmpty()) {
			String symbol = SymbolExtractor.extractDetectedSymbol(line);
			if (symbol != null) {
				return new Resolved(line, SourceMode.CURSOR_WORD, symbol);
			}
		}
		return null;
	}

	private static Resolved fromText(String text, SourceMode baseMode) {
		String trimmed = text.trim();
		String symbol = SymbolExtractor.extractDetectedSymbol(trimmed);
		boolean isSingleSymbol = symbol != null && trimmed.equals(symbol);
		return new Resolved(
				trimmed,
				isSingleSymbol ? SourceMode.SELECTION_SYMBOL : baseMode,
				symbol);
	}

	private static String trim(String s) {
		return s == null ? "" : s.trim();
	}
}
